### PR TITLE
Added transfer for erc20 driver

### PR DESCRIPTION
**Pull request by maaktweluit**
- Added support for `yagna payment transfer` on `--driver erc20`
- Moved logic to `get_next_nonce` to `wallet`

### CODE REVIEW

File: core/payment-driver/erc20/src/driver/cli.rs [hunk 1]
```diff
@@ -83,3 +84,45 @@ If you want to easily acquire some GLM to try Golem on mainnet please use zksync
...
+
+    // Check if there is enough ETH for gas
+    let human_gas_cost = wallet::has_enough_eth_for_gas(&db_tx, network).await?;
+
+    // Everything ok, put the transaction in the queue
+    let tx_id = dao.insert_raw_transaction(db_tx).await;
+
+    let message = format!(
+        "Scheduled {} transfer. from={}, to={}, amount={}, max_gas_cost={} ETH, network={}",
+        &token, &sender, &recipient, &amount, &human_gas_cost, &network
```
Comment (Wiezzel) [inline core/payment-driver/erc20/src/driver/cli.rs:123]: How about using `details` object here? We could then avoid cloning all these values when creating it.

File: core/payment-driver/erc20/src/erc20/wallet.rs [hunk 2]
```diff
@@ -65,8 +65,43 @@ pub async fn fund(dao: &Erc20Dao, address: H160, network: Network) -> Result<(),
...
 }
 
-pub async fn get_network_nonce(address: H160, network: Network) -> Result<U256, GenericError> {
-    ethereum::get_next_nonce(address, network).await
+pub async fn transfer(dao: &Erc20Dao, address: H160, network: Network) -> Result<(), GenericError> {
+    if network == Network::Mainnet {
+        return Err(GenericError::new("Wallet can not be funded on mainnet."));
+    }
+    faucet::request_glm(dao, address, network).await?;
+    Ok(())
```
Comment (Wiezzel) [inline core/payment-driver/erc20/src/erc20/wallet.rs:73]: Looks like you copied `fund` function body here.

File: core/model/src/driver.rs [hunk 3]
```diff
@@ -434,6 +434,19 @@ impl Transfer {
...
+    }
+    pub fn sender(&self) -> String {
+        self.sender.clone()
+    }
+    pub fn to(&self) -> String {
+        self.to.clone()
+    }
+    pub fn network(&self) -> Option<String> {
+        self.network.clone()
+    }
```
[the remark this answers is not in the document]
Comment (maaktweluit) [reply core/model/src/driver.rs:449]: fixed :)

File: core/payment-driver/erc20/src/driver/cli.rs (hunk 1 above)
Comment (maaktweluit) [reply core/payment-driver/erc20/src/driver/cli.rs:123]: good one! updated :)

File: core/model/src/driver.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (maaktweluit) [reply core/model/src/driver.rs]: whoops, indeed forgot to commit that part 😊

File: core/payment-driver/erc20/src/driver/cli.rs [hunk 4]
```diff
@@ -83,3 +84,45 @@ If you want to easily acquire some GLM to try Golem on mainnet please use zksync
...
+    // Everything ok, put the transaction in the queue
+    let tx_id = dao.insert_raw_transaction(db_tx).await;
+
+    let message = format!(
+        "Scheduled {} transfer. details={:?}, max_gas_cost={} ETH, network={}",
+        &token, &details, &human_gas_cost, &network
+    );
+    log::info!("{}", message);
+    log::debug!("tx_id={}", tx_id);
+    Ok(message)
```
Comment (Wiezzel) [inline core/payment-driver/erc20/src/driver/cli.rs:127]: This message is fine for logging but for user output I'd go with etherscan link.

**Review by Wiezzel — APPROVED** [review]
Working fine. My only suggestion is to improve the output message.

File: core/payment-driver/erc20/src/driver/cli.rs (hunk 4 above)
Comment (maaktweluit) [reply core/payment-driver/erc20/src/driver/cli.rs:127]: good catch! made an issue for this: https://golemproject.atlassian.net/browse/PAY-292

**Pull request merged**